### PR TITLE
Shuffle agg

### PR DESCRIPTION
**Pull request by fzhedu**
support shuffle agg including max(), min(), count(), sum().

TODO：

1. avg
2. scalar agg：like select count(*) from table; 
3. distinct


test on ssb
```

mysql> select d_year, max(d_month),sum(d_datekey),min(d_month),count(*) from date group by d_year;
+--------+--------------+----------------+--------------+----------+
| d_year | max(d_month) | sum(d_datekey) | min(d_month) | count(*) |
+--------+--------------+----------------+--------------+----------+
|   1992 | September    |     7290964167 | April        |      366 |
|   1993 | September    |     7274693938 | April        |      365 |
|   1998 | September    |     7272962707 | April        |      364 |
|   1994 | September    |     7278343938 | April        |      365 |
|   1997 | September    |     7289293938 | April        |      365 |
|   1995 | September    |     7281993938 | April        |      365 |
|   1996 | September    |     7305604167 | April        |      366 |
+--------+--------------+----------------+--------------+----------+
7 rows in set (0.11 sec)

mysql> desc select d_year, max(d_month),sum(d_datekey),min(d_month),count(*) from date group by d_year;
+------------------------------------+----------+--------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                 | estRows  | task         | access object | operator info                                                                                                                                                                                                        |
+------------------------------------+----------+--------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Projection_4                       | 8000.00  | root         |               | ssb50.date.d_year, Column#17, Column#18, Column#19, Column#20                                                                                                                                                        |
| └─TableReader_18                   | 8000.00  | root         |               | data:ExchangeSender_17                                                                                                                                                                                               |
|   └─ExchangeSender_17              | 8000.00  | cop[tiflash] |               | ExchangeType: PassThrough                                                                                                                                                                                            |
|     └─HashAgg_16                   | 8000.00  | cop[tiflash] |               | group by:ssb50.date.d_year, funcs:max(Column#26)->Column#17, funcs:sum(Column#27)->Column#18, funcs:min(Column#28)->Column#19, funcs:sum(Column#29)->Column#20, funcs:firstrow(ssb50.date.d_year)->ssb50.date.d_year |
|       └─ExchangeReceiver_15        | 10000.00 | cop[tiflash] |               |                                                                                                                                                                                                                      |
|         └─ExchangeSender_14        | 10000.00 | cop[tiflash] |               | ExchangeType: HashPartition, Hash Cols: ssb50.date.d_year                                                                                                                                                            |
|           └─HashAgg_7              | 8000.00  | cop[tiflash] |               | group by:ssb50.date.d_year, funcs:max(ssb50.date.d_month)->Column#26, funcs:sum(ssb50.date.d_datekey)->Column#27, funcs:min(ssb50.date.d_month)->Column#28, funcs:count(1)->Column#29                                |
|             └─TableFullScan_13     | 10000.00 | cop[tiflash] | table:date    | keep order:false, stats:pseudo                                                                                                                                                                                       |
+------------------------------------+----------+--------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
8 rows in set (0.00 sec)



```

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -1467,14 +1471,50 @@ func (p *PhysicalHashAgg) attach2Task(tasks ...task) task {
 			attachPlan2Task(p, t)
 		}
 	} else if mpp, ok := t.(*mppTask); ok {
-		partialAgg, finalAgg := p.newPartialAggregate(kv.TiFlash)
-		if partialAgg != nil {
-			partialAgg.SetChildren(mpp.p)
-			mpp.p = partialAgg
+		// TODO : && t.count() > float64(p.SCtx().GetSessionVars().BroadcastJoinThresholdCount
+		if p.SCtx().GetSessionVars().AllowMPPExecution && p.tp == plancodec.TypeHashAgg {
```
Comment (hanfei1991) [inline planner/core/task.go]: If we go to this branch, the condition is forever true, so this check is meaning less

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -599,13 +599,14 @@ func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candida
...
-func (ds *DataSource) findBestTask(prop *property.PhysicalProperty, planCounter *PlanCounterTp) (t task, cntPlan int64, err error) {
+func (ds *DataSource) findBestTask(propp *property.PhysicalProperty, planCounter *PlanCounterTp) (t task, cntPlan int64, err error) {
 	// If ds is an inner plan in an IndexJoin, the IndexJoin will generate an inner plan by itself,
 	// and set inner child prop nil, so here we do nothing.
-	if prop == nil {
+	if propp == nil {
 		planCounter.Dec(1)
 		return nil, 1, nil
 	}
+	prop := propp.Clone()
```
Comment (hanfei1991) [inline planner/core/find_best_task.go]: this clone is meaningless

File: planner/core/task.go [hunk 3]
```diff
@@ -1472,38 +1472,36 @@ func (p *PhysicalHashAgg) attach2Task(tasks ...task) task {
...
+		if mpp.partTp == property.HashType {
+			if receiver, ok := mpp.p.(*PhysicalExchangeReceiver); !ok {
+				/// 1-phase agg: when the partition columns can be satisfied, where the plan does not need to enforce Exchange
 				/// only push down the original agg
 				p.self.SetChildren(mpp.p)
 				mpp.p = p.self
 				mpp.addCost(p.GetCost(inputRows, false))
 				return mpp
-			} else if mpp.partTp == property.HashType {
+			} else {
```
Comment (hanfei1991) [inline planner/core/task.go]: I think the logic is reversed:

- If we find a exchanger, it means we need to do 2-phase agg, or don't do it if the gby column's cardinality is low.
- If we don't find a exchanger, it means the task **suffices** the partition requirements. We just need to do 1-phase agg.

File: planner/core/task.go [hunk 4]
```diff
@@ -1467,14 +1471,59 @@ func (p *PhysicalHashAgg) attach2Task(tasks ...task) task {
...
+					p.self.SetChildren(mpp.p)
+					mpp.p = p.self
+					mpp.addCost(p.GetCost(inputRows, false))
+					return mpp
+				} else {
+					/// 2-phase agg: partial + final agg with two types partitions: hash partition or collected partition
+					// TODO: add collected partition agg
+					partialAgg, finalAgg := p.newPartialAggregate(kv.TiFlash)
+					if partialAgg == nil {
+						finalAgg.SetChildren(mpp.p)
```
Comment (fzhedu) [inline planner/core/task.go]: return invalidTask.

File: planner/core/exhaust_physical_plans.go [hunk 5]
```diff
@@ -2223,9 +2233,32 @@ func (la *LogicalAggregation) getHashAggs(prop *property.PhysicalProperty) []Phy
 		taskTypes = []property.TaskType{prop.TaskTp}
 	}
 	for _, taskTp := range taskTypes {
-		agg := NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, TaskTp: taskTp})
-		agg.SetSchema(la.schema.Clone())
-		hashAggs = append(hashAggs, agg)
+		if taskTp == property.MppTaskType && la.checkCanPushDownToMPP() {
```
Comment (fzhedu) [inline planner/core/exhaust_physical_plans.go]: move it to its later.

**Pull request closed without merging**